src/modules/deposit/service.ts

```diff
@@ -34,11 +36,58 @@ export class DepositService {
   constructor(
     private appConfig: AppConfig,
     @InjectRepository(Deposit) private depositRepository: Repository<Deposit>,
+    @InjectRepository(Block) private blockRepository: Repository<Block>,
+    @InjectRepository(SetPoolRebalanceRouteEvent)
+    private setPoolRebalanceRouteRepository: Repository<SetPoolRebalanceRouteEvent>,
     @Inject(CACHE_MANAGER) private cacheManager: Cache,
     private dataSource: DataSource,
     private rewardService: RewardService,
   ) {}
 
+  public async deriveOutputTokenAddress(
+    originChainId: number,
+    inputTokenAddress: string,
+    destinationChainId: number,
+    depositBlockNumber: number,
+    quoteTimestamp: number) {
+    const quoteDatetime = new Date(quoteTimestamp * 1000);
+    const mainnetBlockNumber = originChainId === ChainIds.mainnet
+      ? depositBlockNumber : (await this.blockRepository.findOne({
+        where: {
+          chainId: ChainIds.mainnet,
+          date: LessThanOrEqual(quoteDatetime),
+        },
+        order: { date: "DESC" },
+      })).blockNumber;
```

Comment:
@amateima There's an edge case with this approach. This will fail if you don't have a block matching the criteria on mainnet. Some workarounds I can think of are:
- Using the SDK's `BlockFinder` class. It has a `getBlockForTimestamp` function we can use instead of the Block table.
- Using DefiLlama's API. It has an endpoint that also resolves timestamps to block numbers. See `https://coins.llama.fi/block/{chain}/{timestamp}` docs [here](https://defillama.com/docs/api).
- Not a complete fix but we could also filter by blockNumber only when there's more than one rebalance route for a token.

I think this current logic will work in production as the block table has enough data (and will always do(?)) but I just wanted to flag this in case you consider this is something we should fix before merging.